grpc/src/main/java/io/stargate/grpc/service/MessageHandler.java

```diff
@@ -0,0 +1,455 @@
+import org.apache.cassandra.stargate.exceptions.PreparedQueryNotFoundException;
+import org.apache.cassandra.stargate.exceptions.ReadFailureException;
+import org.apache.cassandra.stargate.exceptions.ReadTimeoutException;
+import org.apache.cassandra.stargate.exceptions.UnavailableException;
+import org.apache.cassandra.stargate.exceptions.WriteFailureException;
+import org.apache.cassandra.stargate.exceptions.WriteTimeoutException;
+
+/**
+ * @param <MessageT> the type of gRPC message being handled.
+ * @param <PreparedT> the persistence object resulting from the preparation of the query(ies).
+ */
+abstract class MessageHandler<MessageT extends GeneratedMessageV3, PreparedT> {
+
+  static final Metadata.Key<QueryOuterClass.Unavailable> UNAVAILABLE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.Unavailable.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteTimeout> WRITE_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadTimeout> READ_TIMEOUT_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadTimeout.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.ReadFailure> READ_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.ReadFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.FunctionFailure> FUNCTION_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.FunctionFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.WriteFailure> WRITE_FAILURE_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.WriteFailure.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.AlreadyExists> ALREADY_EXISTS_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.AlreadyExists.getDefaultInstance());
+  static final Metadata.Key<QueryOuterClass.CasWriteUnknown> CAS_WRITE_UNKNOWN_KEY =
+      ProtoUtils.keyForProto(QueryOuterClass.CasWriteUnknown.getDefaultInstance());
+
+  protected static final ConsistencyLevel DEFAULT_TRACING_CONSISTENCY = ConsistencyLevel.ONE;
+
+  protected final MessageT message;
+  protected final Connection connection;
+  private final Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache;
+  protected final Persistence persistence;
+  private final StreamObserver<Response> responseObserver;
+
+  protected MessageHandler(
+      MessageT message,
+      Connection connection,
+      Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache,
+      Persistence persistence,
+      StreamObserver<Response> responseObserver) {
+    this.message = message;
+    this.connection = connection;
+    this.preparedCache = preparedCache;
+    this.persistence = persistence;
+    this.responseObserver = responseObserver;
+  }
+
+  void handle() {
+    CompletionStage<Result> resultFuture =
+        CompletableFuture.<Void>completedFuture(null)
+            .thenApply(this::invokeValidate)
```

Comment:
This is a bit ugly, but it allows us to handle all errors with a single `handleException` at the end of the future chain.

Reply:
I think I'd prefer something like the following even though we have duplicate calls to `handleException()`:

```
    try {
      validate();

      CompletionStage<Result> resultCompletionStage = prepare(false)
          .thenCompose(this::executePrepared);

      resultCompletionStage = handleUnprepared(resultCompletionStage);
      resultCompletionStage
          .thenApply(this::buildResponse)
          .thenCompose(this::executeTracingQueryIfNeeded)
          .whenComplete(
              (response, error) -> {
                if (error != null) {
                  handleException(error);
                } else {
                  setSuccess(response);
                }
              });
    } catch (Throwable t) {
      handleException(t);
    }
```

1) It's simpler; it doesn't require a temporary future and we don't have to have the method `invokeValidate()`.
2) It's a bit safer because if any of the async setup code throws it'll be handled and won't hang the request. The code might be exception safe now until someone adds something in the future that's not.

Reply:
OK, I was a bit on the fence about this, I don't mind changing it back.